Library/SPIClient.m

```diff
@@ -2109,7 +2107,7 @@ - (NSString *)validatePosId:(NSString *)posId {
 }
 
 - (void)validateEftposAddress:(NSString *)eftposAddress {
-    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:regexItemsForEftposAddress options:NSRegularExpressionCaseInsensitive error:nil];
+    regex = [NSRegularExpression regularExpressionWithPattern:regexItemsForEftposAddress options:NSRegularExpressionCaseInsensitive error:nil];
```

Comment:
First of all, `regex` is not a good name for an instance variable. Regex for what?

Secondly, what was the point of making it an instance variable if you're re-instantiating it every time `validateEftposAddress` gets called? You're meant to instantiate it once for the instance of the class and reuse it...

Reply:
**`NSRegularExpression *regexForPosId = [NSRegularExpression regularExpressionWithPattern:@"^[a-zA-Z0-9 ]*$" options:NSRegularExpressionCaseInsensitive error:nil];`**

I got the error like below when it is instantiated once.

**initializer element is not a compile-time constant**

The below explanation is on https://developer.apple.com/documentation/foundation/nsregularexpression?language=objc

> Concurrency and Thread Safety
> NSRegularExpression is designed to be immutable and thread-safe so that a single instance can be used in matching operations on multiple threads at once. However, the string on which it is operating should not be mutated during the course of a matching operation, whether from another thread or from within the Block used in the iteration.

Please show me what is the right way to do. Thanks.

Reply:
That's because you made it static, which is not allowed since `NSRegularExpression` is not constant (like `NSString` or `NSInteger`). An existing example of a private instance variable is `txLock`, try declaring and instantiating this `posIdRegex` variable in a similar way to that.

Reply:
it's fixed